### PR TITLE
Improve the performance of currency

### PR DESCRIPTION
**Pull request by Sothan**
- Add some changes to PR #949 
- Add few specs into CurrencyModel
- Improve about 21% in init
- Improve about 40% in parsing

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.NumberWithUnit/Extractors/NumberWithUnitExtractor.cs [hunk 1]
```diff
@@ -132,209 +146,154 @@ public bool ValidateUnit(string source)
...
-                    var match = ambiguousMultiplierRegex.Matches(number.Text);
-                    if (match.Count == 1)
-                    {
-                        var newLength = number.Text.Length - match[0].Length;
-                        number.Text = number.Text.Substring(0, newLength);
-                        number.Length = newLength;
-                    }
-                }
-            }
+            var nonUnitMatch = this.config.PmNonUnitRegex.Matches(source);
```
Comment (wgx998877) [inline .NET/Microsoft.Recognizers.Text.NumberWithUnit/Extractors/NumberWithUnitExtractor.cs]: set this null as default, and do the matches when use it.

File: .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                return tokens;
+            }
+
+            bool inToken = false;
+            int tokenStart = 0;
+            var chars = input.ToCharArray();
+            for (int i = 0; i < chars.Length; i++)
+            {
+                var c = chars[i];
+                if (!char.IsLetter(c) || IsCjk(c))
```
Comment (wgx998877) [inline .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs]: Please add more unit with symbol cases in the test files?

File: .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                var c = chars[i];
+                if (char.IsWhiteSpace(c))
+                {
+                    if (inToken)
+                    {
+                        tokens.Add(new Token(tokenStart, i - tokenStart, input.Substring(tokenStart, i - tokenStart)));
+                        inToken = false;
+                    }
+                }
+                // Let 's$', 'us$' etc. as a token
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs]: Nit: Comments should be to the side of the if or inside it.

File: .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+            if (inToken)
+            {
+                tokens.Add(new Token(tokenStart, chars.Length - tokenStart, input.Substring(tokenStart, chars.Length - tokenStart)));
+            }
+
+            return tokens;
+        }
+
+        private bool CaseByContext(char curChar, char preChar)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs]: I think this method can get a better name. It's not clear what "context" means here.

File: .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            // Split tokens like '200ml' to two separate tokens, i.e., '200' and 'ml'
+            if ((char.IsLetter(curChar) && char.IsDigit(preChar)) || (char.IsDigit(curChar) && char.IsLetter(preChar)))
+            {
+                return true;
+            }
+
+            // For cases like '$100' or '$100' shouldn't be a token
+            if ((char.IsDigit(curChar) && preChar == '$') || (curChar == '$' && char.IsDigit(preChar)))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs]: This doesn't seem to make sense. Isn't $ a prefix and 100 a number, in a similar way to 200 as number and ml as suffix above?

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text/Matcher/NumberWithUnitTokenizer.cs]: Seems only the comment is wrong.

**Pull request merged**